tls/s2n_handshake.c

```diff
@@ -40,14 +40,15 @@ int s2n_handshake_write_header(struct s2n_stuffer *out, uint8_t message_type)
 
 int s2n_handshake_finish_header(struct s2n_stuffer *out)
 {
-    uint16_t length = s2n_stuffer_data_available(out);
+    uint32_t length = s2n_stuffer_data_available(out);
     S2N_ERROR_IF(length < TLS_HANDSHAKE_HEADER_LENGTH, S2N_ERR_SIZE_MISMATCH);
 
-    uint16_t payload = length - TLS_HANDSHAKE_HEADER_LENGTH;
+    uint32_t payload = length - TLS_HANDSHAKE_HEADER_LENGTH;
 
     /* Write the message header */
     POSIX_GUARD(s2n_stuffer_rewrite(out));
     POSIX_GUARD(s2n_stuffer_skip_write(out, 1));
+    POSIX_ENSURE(payload <= UINT24_MAX, S2N_ERR_INTEGER_OVERFLOW);
     POSIX_GUARD(s2n_stuffer_write_uint24(out, payload));
```

Comment:
Would it make more sense to add the <= UINT24_MAX check to s2n_stuffer_write_uint24? It should be true for every call to s2n_stuffer_write_uint24, not just this one. It also wouldn't make a lot of sense for UINT24_MAX to live in s2n_stuffer.h if it's only used here.